### ISSUE TITLE
Queueing/Throttling

### ISSUE THREAD

**Issue by prettydiff** [whatwg/websockets#33]
According to MDN the send method enqueues data of supported types for transfer, however the data is not queued across a variety of user agents.  When data is not queued a transmission payload is abandoned from transfer if in conflict with a more recent transmission. 
 This is because a websocket is a single socket and RFC 6455 does not permit the interlacing of data frames.  With some additions to websockets in the browser developers can write their own userland queue.

According to RFC 6455 a transmission is complete when both the first bit of the first byte of a payload fragment is 1 and at the byte size specified by the payload size bytes of the frame header.  I can account for this in Node.js and write my own queue implementation at both the sender and receiver to ensure that transmissions do not conflict or interlace and also account for instances when frame headers are separated from frame bodies or if two separate frames are concatenated into a single transmission.  

The browser does not provide me the capability to read frame headers on incoming transmissions, the message event, so I cannot solve for this problem according to the definitions provided in RFC 6455.  My solution to this problem is to two-fold:

1. Add a new readyState value - **4 - transmitting**
2. Add a new event - **finish**

The ready state value indicates the channel is open and actively writing data to the socket.  The send method changes a socket's ready state from 0, open, to 4, transmitting.  When the transmission completes the ready state returns to 0 and the event *finish* executes.

This is beneficial for circumventing long polling.  If, for example, a given socket were used for a media transmission that socket may be sending binary data for an extended and undetermined length of time.  It would frustrate the user if that transmission fails because the socket executes another send method on a user agent that isn't properly enqueueing transmissions.  The ideal user experience is for the application to know the current state of the socket so that it may open an additional socket for trivial data transmissions.  Even on user agents that are properly enqueue transmissions it may be a better user experience to open a parallel socket as the needs of the user demand, but there is not an automated means of knowing when such actions are appropriate.

**Comment by ricea** [comment]
> According to MDN the send method enqueues data of supported types for transfer, however the data is not queued across a variety of user agents.

I'm not sure what this sentence means. The [standard text](https://websockets.spec.whatwg.org/#dom-websocket-send) doesn't explicitly say that the data is queued, but it's very clear that either the data is sent or the connection is closed. The standard text may be a little bit unclear, but queuing is the only reasonable implementation.

> When data is not queued a transmission payload is abandoned from transfer if in conflict with a more recent transmission.

I don't understand this either. Messages are sent in strict order.

**Comment by prettydiff** [comment]
MDN reference stating the send method enqueues data: https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/send

Some scenarios:

If a send method is executed on a web socket and then another send method is executed on that socket a few lines down within the same function there will be a collision.  The JavaScript language will execute the instructions faster than network hardware can process these instructions into messages over the network.  When this collision occurs the first payload will be abandoned and only the second payload will be sent.  I have verified this on both Firefox and Chrome.

If a web socket is sending media and while that media is sending any use of a send method upon that socket will halt the media transfer by imposing a collision.

---

These collisions, or even the potential thereof, is the biggest limitation of web sockets.  It is a great loss of network fidelity that is easily reproducible.  The point of this issue is that with the proper tools available this problem can be easily managed and mitigated.  I have already solved this issue in a web socket implementation I wrote for Node.js.  It would benefit everybody if the proper signaling were equally available in the browser.
